File: ScriptsToProcess/Facade_Parser.py
"""
FACADE JSON PARSER (V2.1 - DUAL LAYER EDITION)
================================================================================
INPUTS:
    Facade_JSON     (str)  [Item Access]
    Filter_Building (str)  [List Access]
    Filter_Tower    (str)  [List Access]
    Filter_Program  (str)  [List Access]
    Filter_Level    (int)  [List Access]
    ExactMatch      (bool) [Item Access]

OUTPUTS:
    BoundsExt       (Curve) [Open Exterior Physical Lines]
    BoundsClosed    (Curve) [Closed Master Polygon for orientation]
    Heights         (float) [Floor-to-floor height]
    Programs        (str)   [Tags/Labels]
================================================================================
"""
import json
import Rhino.Geometry as rg
import Grasshopper as gh
from Grasshopper.Kernel.Data import GH_Path
import System
import fnmatch
import time

def update_ui(msg):
    _ghenv = globals().get('ghenv')
    _comp = globals().get('component')
    if _ghenv: _ghenv.Component.Message = msg
    elif _comp: _comp.Message = msg

_ghenv = globals().get('ghenv')
if _ghenv:
    _ghenv.Component.Name = "Facade JSON Parser"
    _ghenv.Component.NickName = "Facade_Parser"

# NEW: Deserializes without forcing the curves closed!
def deserialize_open_curves(segments_data):
    if not segments_data: return []
    crvs = []
    for seg in segments_data:
        stype = seg.get("type")
        if stype == "Line": crvs.append(rg.LineCurve(rg.Point3d(*seg["start"]), rg.Point3d(*seg["end"])))
        elif stype == "Arc": crvs.append(rg.ArcCurve(rg.Arc(rg.Point3d(*seg["start"]), rg.Point3d(*seg["mid"]), rg.Point3d(*seg["end"]))))
        elif stype == "Polyline": crvs.append(rg.PolylineCurve([rg.Point3d(*p) for p in seg["points"]]))
    if not crvs: return []
    if len(crvs) == 1: return [crvs[0]]
    
    joined = rg.Curve.JoinCurves(crvs, 0.01)
    return list(joined) if joined else crvs

def is_match(target_name, filter_list, exact_mode):
    if not filter_list: return True 
    target = str(target_name).strip().upper()
    for f in filter_list:
        pattern = str(f).strip().upper()
        if exact_mode:
            if target == pattern: return True
        else:
            if '*' in pattern or '?' in pattern:
                if fnmatch.fnmatch(target, pattern): return True
            elif pattern in target:
                return True
    return False

def query_facades():
    exec_start = time.perf_counter()
    json_in = globals().get('Facade_JSON', "")
    exact_toggle = globals().get('ExactMatch')
    exact_toggle = bool(exact_toggle) if exact_toggle is not None else False
    
    raw_bldg = globals().get('Filter_Building')
    f_bldg = [b for b in raw_bldg if b] if raw_bldg else []
    
    raw_tower = globals().get('Filter_Tower')
    f_tower = [t for t in raw_tower if t] if raw_tower else []
    
    raw_prog = globals().get('Filter_Program')
    f_prog = [p for p in raw_prog if p] if raw_prog else []
    
    raw_level = globals().get('Filter_Level')
    f_level = []
    if raw_level:
        for lvl in raw_level:
            if lvl is not None:
                try: f_level.append(int(lvl))
                except: pass
    
    # Initialize output DataTrees
    out_bounds_ext = gh.DataTree[System.Object]()
    out_bounds_closed = gh.DataTree[System.Object]()
    out_heights = gh.DataTree[System.Object]()
    out_programs = gh.DataTree[System.Object]()
    
    if not json_in:
        update_ui("FACADE PARSER\nTime: 0.0 ms\n---\nAwaiting Data")
        return out_bounds_ext, out_bounds_closed, out_heights, out_programs
        
    try:
        data = json.loads(json_in)
        bldg_index = 0
        match_count = 0
        
        for bldg_name, prog_dict in data.items():
            if not is_match(bldg_name, f_bldg, exact_toggle): continue
            prog_index = 0
            
            for prog_name, floors in prog_dict.items():
                if not is_match(prog_name, f_prog, exact_toggle): continue
                path = GH_Path(bldg_index, prog_index)
                
                for floor in floors:
                    tower_id = floor.get("tower_id", "Unknown")
                    if f_level and floor.get("floor_index", -1) not in f_level: continue
                    if not is_match(tower_id, f_tower, exact_toggle): continue
                    
                    # Extract True Z Elevation
                    z_offset = rg.Transform.Translation(0, 0, floor.get("true_z", 0.0))
                    
                    # 1. Evaluate Closed Master Polygon
                    closed_crvs = deserialize_open_curves(floor.get("BoundsClosed", []))
                    closed_crv = None
                    if closed_crvs:
                        closed_crv = closed_crvs[0]
                        if not closed_crv.IsClosed: closed_crv.MakeClosed(0.01)
                        closed_crv.Transform(z_offset)
                    
                    # 2. Evaluate Open Exterior Line Segments
                    ext_crvs = deserialize_open_curves(floor.get("BoundsExt", []))
                    
                    label = f"{bldg_name} | {tower_id} | {prog_name} - Lvl {floor.get('floor_index', '?')}"
                    
                    # 3. Parallel Assignment (matches 1 closed poly to multiple open lines)
                    for ext_crv in ext_crvs:
                        ext_crv.Transform(z_offset)
                        
                        out_bounds_ext.Add(ext_crv, path)
                        out_bounds_closed.Add(closed_crv, path)
                        out_heights.Add(floor["height"], path)
                        out_programs.Add(label, path)
                        match_count += 1
                        
                prog_index += 1
            bldg_index += 1
            
        exec_time = (time.perf_counter() - exec_start) * 1000
        search_mode = "Exact" if exact_toggle else "Flexible"
        update_ui(f"FACADE PARSER\nTime: {exec_time:.0f} ms\n---\nReturned: {match_count}\nMode: {search_mode}")
        return out_bounds_ext, out_bounds_closed, out_heights, out_programs
        
    except Exception as e:
        update_ui("JSON Parse Error:\n" + str(e))
        return out_bounds_ext, out_bounds_closed, out_heights, out_programs

# Direct the function outputs to the Grasshopper component output ports
bnds_ext, bnds_closed, hts, progs = query_facades()

BoundsExt = bnds_ext
BoundsClosed = bnds_closed
Heights = hts
Programs = progs

File: Components/FacadeParser.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Text.RegularExpressions;
using Newtonsoft.Json.Linq;
using Rhino.Geometry;
using Grasshopper.Kernel;
using Grasshopper.Kernel.Data;
using Grasshopper.Kernel.Types;

namespace Enzyme.Components
{
    public class FacadeParserComponent : GH_Component
    {
        public FacadeParserComponent()
          : base("Facade JSON Parser", "Facade_Parser",
              "Parses Facade JSON data into Curves, Heights, and Programs.",
              Enzyme.Utils.TabInfo.TabName, "Masterplan (Beta)")
        {
        }

                public override void AddedToDocument(GH_Document document)
        {
            base.AddedToDocument(document);
            if (this.Attributes == null) this.CreateAttributes();

            bool hasSources = false;
            foreach (var param in this.Params.Input)
                if (param.SourceCount > 0) { hasSources = true; break; }

            if (!hasSources)
            {
                Enzyme.Utils.AutoWireHelper.WireToggle(this, document, 5, false, 210, 0);
                Enzyme.Utils.AutoWireHelper.WireOutputParam(this, document, 0, "curve", 220, -45);
                Enzyme.Utils.AutoWireHelper.WireOutputParam(this, document, 1, "curve", 220, 0);
                Enzyme.Utils.AutoWireHelper.WireOutputPanel(this, document, 3, 220, 34, 180, 22);
            }
        }

        protected override void RegisterInputParams(GH_Component.GH_InputParamManager pManager)
        {
            pManager.AddTextParameter("Facade_JSON", "Facade_JSON", "JSON string representing facade data", GH_ParamAccess.item);
            pManager.AddTextParameter("Filter_Building", "Filter_Building", "Filter by building name", GH_ParamAccess.list);
            pManager.AddTextParameter("Filter_Tower", "Filter_Tower", "Filter by tower ID", GH_ParamAccess.list);
            pManager.AddTextParameter("Filter_Program", "Filter_Program", "Filter by program name", GH_ParamAccess.list);
            pManager.AddIntegerParameter("Filter_Level", "Filter_Level", "Filter by level index", GH_ParamAccess.list);
            pManager.AddBooleanParameter("ExactMatch", "ExactMatch", "Require exact match for filters", GH_ParamAccess.item, false);

            pManager[1].Optional = true;
            pManager[2].Optional = true;
            pManager[3].Optional = true;
            pManager[4].Optional = true;
        }

        protected override void RegisterOutputParams(GH_Component.GH_OutputParamManager pManager)
        {
            pManager.AddCurveParameter("BoundsExt", "BoundsExt", "Open Exterior Physical Lines", GH_ParamAccess.tree);
            pManager.AddCurveParameter("BoundsClosed", "BoundsClosed", "Closed Master Polygon for orientation", GH_ParamAccess.tree);
            pManager.AddNumberParameter("Heights", "Heights", "Floor-to-floor height", GH_ParamAccess.tree);
            pManager.AddTextParameter("Programs", "Programs", "Tags/Labels", GH_ParamAccess.tree);
        }

        protected override void SolveInstance(IGH_DataAccess DA)
        {
            Stopwatch sw = Stopwatch.StartNew();

            string jsonIn = null;
            if (!DA.GetData(0, ref jsonIn) || string.IsNullOrWhiteSpace(jsonIn))
            {
                this.Message = this.NickName + "\nTime: 0.0 ms\n---\nAwaiting Data";
                return;
            }

            List<string> fBldg = new List<string>();
            DA.GetDataList(1, fBldg);
            fBldg = fBldg.Where(x => !string.IsNullOrWhiteSpace(x)).ToList();

            List<string> fTower = new List<string>();
            DA.GetDataList(2, fTower);
            fTower = fTower.Where(x => !string.IsNullOrWhiteSpace(x)).ToList();

            List<string> fProg = new List<string>();
            DA.GetDataList(3, fProg);
            fProg = fProg.Where(x => !string.IsNullOrWhiteSpace(x)).ToList();

            List<int> fLevel = new List<int>();
            DA.GetDataList(4, fLevel);

            bool exactToggle = false;
            DA.GetData(5, ref exactToggle);

            GH_Structure<GH_Curve> outBoundsExt = new GH_Structure<GH_Curve>();
            GH_Structure<GH_Curve> outBoundsClosed = new GH_Structure<GH_Curve>();
            GH_Structure<GH_Number> outHeights = new GH_Structure<GH_Number>();
            GH_Structure<GH_String> outPrograms = new GH_Structure<GH_String>();

            int matchCount = 0;

            try
            {
                JObject data = JObject.Parse(jsonIn);
                int bldgIndex = 0;

                foreach (var bldgKvp in data)
                {
                    string bldgName = bldgKvp.Key;
                    if (!IsMatch(bldgName, fBldg, exactToggle)) continue;
                    
                    JObject progDict = bldgKvp.Value as JObject;
                    if (progDict == null) continue;

                    int progIndex = 0;

                    foreach (var progKvp in progDict)
                    {
                        string progName = progKvp.Key;
                        if (!IsMatch(progName, fProg, exactToggle)) continue;
                        
                        GH_Path path = new GH_Path(bldgIndex, progIndex);
                        JArray floors = progKvp.Value as JArray;
                        if (floors == null) continue;

                        foreach (JObject floor in floors)
                        {
                            string towerId = floor["tower_id"]?.ToString() ?? "Unknown";
                            
                            // Treat missing floor_index as -1 matching the python code
                            int floorIndex = -1;
                            if (floor["floor_index"] != null && floor["floor_index"].Type != JTokenType.Null)
                            {
                                int.TryParse(floor["floor_index"].ToString(), out floorIndex);
                            }

                            if (fLevel.Count > 0 && !fLevel.Contains(floorIndex)) continue;
                            if (!IsMatch(towerId, fTower, exactToggle)) continue;

                            double trueZ = floor["true_z"]?.ToObject<double>() ?? 0.0;
                            Transform zOffset = Transform.Translation(0, 0, trueZ);

                            List<Curve> closedCrvs = DeserializeOpenCurves(floor["BoundsClosed"] as JArray);
                            Curve closedCrv = null;
                            if (closedCrvs.Count > 0)
                            {
                                closedCrv = closedCrvs[0];
                                if (!closedCrv.IsClosed)
                                {
                                    closedCrv.MakeClosed(0.01);
                                }
                                closedCrv.Transform(zOffset);
                            }

                            List<Curve> extCrvs = DeserializeOpenCurves(floor["BoundsExt"] as JArray);
                            
                            string floorLabel = floor["floor_index"]?.ToString() ?? "?";
                            string label = $"{bldgName} | {towerId} | {progName} - Lvl {floorLabel}";
                            double height = floor["height"]?.ToObject<double>() ?? 0.0;

                            foreach (Curve extCrv in extCrvs)
                            {
                                extCrv.Transform(zOffset);

                                outBoundsExt.Append(new GH_Curve(extCrv), path);
                                outBoundsClosed.Append(new GH_Curve(closedCrv), path);
                                outHeights.Append(new GH_Number(height), path);
                                outPrograms.Append(new GH_String(label), path);
                                matchCount++;
                            }
                        }
                        progIndex++;
                    }
                    bldgIndex++;
                }

                sw.Stop();
                string searchMode = exactToggle ? "Exact" : "Flexible";
                this.Message = this.NickName + $"\nTime: {sw.ElapsedMilliseconds} ms\n---\nReturned: {matchCount}\nMode: {searchMode}";
            }
            catch (Exception ex)
            {
                this.Message = "JSON Parse Error:\n" + ex.Message;
            }

            DA.SetDataTree(0, outBoundsExt);
            DA.SetDataTree(1, outBoundsClosed);
            DA.SetDataTree(2, outHeights);
            DA.SetDataTree(3, outPrograms);
        }

        private bool IsMatch(string targetName, List<string> filterList, bool exactMode)
        {
            if (filterList == null || filterList.Count == 0) return true;
            string target = (targetName ?? "").Trim().ToUpperInvariant();
            
            foreach (string f in filterList)
            {
                string pattern = (f ?? "").Trim().ToUpperInvariant();
                if (exactMode)
                {
                    if (target == pattern) return true;
                }
                else
                {
                    if (pattern.Contains("*") || pattern.Contains("?"))
                    {
                        string regexPattern = "^" + Regex.Escape(pattern).Replace("\\*", ".*").Replace("\\?", ".") + "$";
                        if (Regex.IsMatch(target, regexPattern)) return true;
                    }
                    else if (target.Contains(pattern))
                    {
                        return true;
                    }
                }
            }
            return false;
        }

        private List<Curve> DeserializeOpenCurves(JArray segmentsData)
        {
            if (segmentsData == null || segmentsData.Count == 0) return new List<Curve>();
            
            List<Curve> crvs = new List<Curve>();
            foreach (JObject seg in segmentsData)
            {
                string stype = seg["type"]?.ToString();
                if (stype == "Line")
                {
                    JArray start = seg["start"] as JArray;
                    JArray end = seg["end"] as JArray;
                    crvs.Add(new LineCurve(new Point3d((double)start[0], (double)start[1], (double)start[2]), 
                                           new Point3d((double)end[0], (double)end[1], (double)end[2])));
                }
                else if (stype == "Arc")
                {
                    JArray start = seg["start"] as JArray;
                    JArray mid = seg["mid"] as JArray;
                    JArray end = seg["end"] as JArray;
                    crvs.Add(new ArcCurve(new Arc(new Point3d((double)start[0], (double)start[1], (double)start[2]),
                                                  new Point3d((double)mid[0], (double)mid[1], (double)mid[2]),
                                                  new Point3d((double)end[0], (double)end[1], (double)end[2]))));
                }
                else if (stype == "Polyline")
                {
                    JArray points = seg["points"] as JArray;
                    List<Point3d> pts = new List<Point3d>();
                    foreach (JArray p in points)
                    {
                        pts.Add(new Point3d((double)p[0], (double)p[1], (double)p[2]));
                    }
                    crvs.Add(new PolylineCurve(pts));
                }
            }

            if (crvs.Count == 0) return new List<Curve>();
            if (crvs.Count == 1) return crvs;

            Curve[] joined = Curve.JoinCurves(crvs, 0.01);
            if (joined != null && joined.Length > 0)
            {
                return joined.ToList();
            }
            return crvs;
        }

        public override GH_Exposure Exposure => GH_Exposure.secondary;

        protected override Bitmap Icon => IconLoader.Load("Facade_Parser.png");

        public override Guid ComponentGuid => new Guid("7F0429C2-82DA-4EAA-9076-E547BA93EAA4");
    }
}
